File: PinDuino/src/SolenoidAction.cpp
/*
 * SolenoidAction.cpp
 *
 *  Created on: Mar 1, 2015
 *      Author: peter
 */

#include "SolenoidAction.h"

SolenoidAction::SolenoidAction() : enabled(1), solenoidIndex(5), attack(65000), sustain(0) {
}

SolenoidAction::SolenoidAction(bool enabled, uint8_t solenoidIndex, uint16_t attack, uint8_t sustain) : enabled(enabled), solenoidIndex(solenoidIndex), attack(attack), sustain(sustain) {
}

void SolenoidAction::trigger() {
	if(attack == 0) {
		solenoids[solenoidIndex].release();
	} else {
		solenoids[solenoidIndex].trigger(attack);
	}
}

File: kingpyn/kingpyn/SolenoidAction.py
'''
Created on May 28, 2015

@author: peter
'''

class SolenoidAction(object):
    '''
    classdocs
    '''


    def __init__(self, enabled, driverId, attack, sustain):
        '''
        Constructor
        '''
        self.enabled = enabled
        self.driverId = driverId
        self.attack = attack
        self.sustain = sustain
        
    def __eq__(self, other):
        if isinstance(other, SolenoidAction):
            return self.enabled == other.enabled and self.driverId == other.driverId and self.attack == other.attack and self.sustain == other.sustain
        return NotImplemented
    
    def __hash__(self):
        return hash(self.enabled) ^ hash(self.driverId) ^ hash(self.attack) ^ hash(self.sustain)
    
    def toByteArray(self):
        return [self.enabled | self.driverId << 1,
                self.sustain,
                self.attack & 255,
                self.attack >> 8 ]
        
    @staticmethod
    def fromByteArray(byteArray):
        enabled = byteArray[0] & 0x01
        driverId = byteArray[0] >> 1
        attack = byteArray[2] + (byteArray[3] << 8)
        sustain = byteArray[1]
        return SolenoidAction(enabled, driverId, attack, sustain)
    
    def __str__(self):
        return "SolenoidAction({}, {}, {}, {})".format(self.enabled, self.driverId, self.attack, self.sustain)
    
    def __repr__(self):
        return self.__str__()